module/zfs/vdev_raidz.c

```diff
@@ -343,47 +608,30 @@ vdev_raidz_map_alloc(zio_t *zio, uint64_t ashift, uint64_t dcols,
 	}
 
 	ASSERT3U(acols, <=, scols);
-
-	rr = kmem_alloc(offsetof(raidz_row_t, rr_col[scols]), KM_SLEEP);
+	rr = vdev_raidz_row_alloc(scols);
 	rm->rm_row[0] = rr;
-
 	rr->rr_cols = acols;
-	rr->rr_scols = scols;
 	rr->rr_bigcols = bc;
-	rr->rr_missingdata = 0;
-	rr->rr_missingparity = 0;
 	rr->rr_firstdatacol = nparity;
-	rr->rr_abd_empty = NULL;
-	rr->rr_nempty = 0;
 #ifdef ZFS_DEBUG
 	rr->rr_offset = zio->io_offset;
 	rr->rr_size = zio->io_size;
 #endif
 
-	asize = 0;
+	uint64_t asize = 0;
 
-	for (c = 0; c < scols; c++) {
+	for (uint64_t c = 0; c < scols; c++) {
 		raidz_col_t *rc = &rr->rr_col[c];
-		col = f + c;
-		coff = o;
+		uint64_t col = f + c;
+		uint64_t coff = o;
 		if (col >= dcols) {
 			col -= dcols;
 			coff += 1ULL << ashift;
 		}
 		rc->rc_devidx = col;
 		rc->rc_offset = coff;
-		rc->rc_abd = NULL;
-		rc->rc_orig_data = NULL;
-		rc->rc_error = 0;
-		rc->rc_tried = 0;
-		rc->rc_skipped = 0;
-		rc->rc_force_repair = 0;
-		rc->rc_allow_repair = 1;
-		rc->rc_need_orig_restore = B_FALSE;
 
-		if (c >= acols)
-			rc->rc_size = 0;
-		else if (c < bc)
+		if (c < bc)
 			rc->rc_size = (q + 1) << ashift;
 		else
 			rc->rc_size = q << ashift;
```

Comment:
Good catch.  It seems like we should still be setting `rc_size=0` for the pad sectors, so that we know to issue the optional writes.  Maybe we should see if we're actually issuing the optional writes, and if not then try setting rc_size=0 here as we used to.

Reply:
Changed to restore `rc_size=0` and have been testing with that change.